### PR TITLE
[BUG] Series.diff was always setting the dtype to object

### LINKED ISSUE #30889: Series lose nullable integer dtype after call to .diff()
c.f. #28285

#### Code Sample, a copy-pastable example if possible

```python
pd.Series([1, 2, 3], dtype="Int16").diff()
```
```
0    NaN
1      1
2      1
dtype: object
```

#### Problem description

Output dtype becomes `object`.

#### Expected Output

Output dtype remains `Int16Dtype`.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.6.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.15.0-42-generic
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 0.25.2
numpy            : 1.17.4
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 19.3.1
setuptools       : 44.0.0.post20200102
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.10.3
IPython          : 7.11.1
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : 3.1.2
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
s3fs             : None
scipy            : 1.4.1
sqlalchemy       : None
tables           : None
xarray           : 0.14.1
xlrd             : None
xlwt             : None
xlsxwriter       : None

</details>

### PR DESCRIPTION
**Pull request by MarcoGorelli**
- [x] closes #30889
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/series.py [hunk 1]
```diff
@@ -2301,6 +2301,19 @@ def diff(self, periods=1):
         dtype: float64
         """
         result = algorithms.diff(com.values_from_object(self), periods)
+        if self.dtype in [
+            "Int8",
```
Comment (jbrockmendel) [inline pandas/core/series.py]: it looks like the newer version of numpy is raising on this comparison

File: pandas/core/series.py [hunk 2]
```diff
@@ -2301,6 +2301,19 @@ def diff(self, periods=1):
...
+            "Int16",
+            "Int32",
+            "Int64",
+            "UInt8",
+            "UInt16",
+            "UInt32",
+            "UInt64",
+        ]:
+            return self._constructor(
+                result, index=self.index, dtype=self.dtype
```
Comment (WillAyd) [inline pandas/core/series.py]: What is the downside of always passing `dtype=self.dtype`?

Comment (MarcoGorelli) [reply pandas/core/series.py]: @WillAyd what happens if you do 
```
s = pd.Series([True, False, False]).diff()
```
?

The output will contain a NaN, and if I pass `dtype=self.dtype` in lines 2314-2315 then the resulting series will be of type bool.

It that's what you want, that's fine, but I'll have to change some of the existing tests in pandas/tests/series/methods/test_diff.py

File: pandas/tests/series/methods/test_diff.py [hunk 3]
```diff
@@ -3,6 +3,7 @@
 
 from pandas import Series, TimedeltaIndex, date_range
 import pandas._testing as tm
```
Comment (jreback) [inline pandas/tests/series/methods/test_diff.py]: this is a fixture you don’t need to import 

in fact we should make a code check that bans this (pls make an issue)

File: pandas/core/series.py [hunk 4]
```diff
@@ -2301,6 +2301,19 @@ def diff(self, periods=1):
         dtype: float64
         """
         result = algorithms.diff(com.values_from_object(self), periods)
+        if self.dtype in [
```
Comment (jreback) [inline pandas/core/series.py]: this not a good pattern

if always passing the dtype is not working
then you can use a is_* method here

explicitly listing dtypes is a no no

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by MarcoGorelli** [comment]
@jreback @jbrockmendel @WillAyd thanks for your quick reviews, I've updated the PR accordingly

File: doc/source/whatsnew/v1.0.0.rst [hunk 5]
```diff
@@ -1146,6 +1146,7 @@ ExtensionArray
 - Bug in :class:`arrays.PandasArray` when setting a scalar string (:issue:`28118`, :issue:`28150`).
 - Bug where nullable integers could not be compared to strings (:issue:`28930`)
 - Bug where :class:`DataFrame` constructor raised ``ValueError`` with list-like data and ``dtype`` specified (:issue:`30280`)
+- Bug in :meth:`Series.diff` was always setting the ``dtype`` to ``Object`` (:issue:`30889`)
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v1.0.0.rst]:
```suggestion
- Bug in :meth:`Series.diff` was always setting the ``dtype`` to ``object`` (:issue:`30889`)
```

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks for the PR!

File: pandas/core/series.py [hunk 6]
```diff
@@ -2301,6 +2302,12 @@ def diff(self, periods=1):
         dtype: float64
         """
         result = algorithms.diff(com.values_from_object(self), periods)
+        if is_extension_array_dtype(self.dtype) and not is_datetime64tz_dtype(
+            self.dtype
+        ):
+            return self._constructor(
+                result, index=self.index, dtype=self.dtype
```
[the remark this answers is not in the document]
Comment (MarcoGorelli) [reply pandas/core/series.py]: Do you mean, with a try-except block?

I can do that, but then I need to update the expected output in `pandas/tests/series/methods/test_diff.py::TestSeriesDiff::test_diff_bool`, which, for
```
pd.Series([False, True, True, False, False]).diff()
```
 is currently
```
0      NaN
1     True
2    False
3     True
4    False
dtype: object
```
. If I pass set the dtype in `Series.diff`, then the result will be
```
result
0     True
1     True
2    False
3     True
4    False
dtype: bool
```

Is that what you'd like it to be instead?

[the remark this answers is not in the document]
Comment (TomAugspurger) [reply pandas/core/series.py]: No, I don't think we want that output. That's a consequence of `bool(np.nan)`.

**Review by TomAugspurger — COMMENTED** [review]
Isn't the correct fix to to dispatch `diff` to the EA, rather than trying to fix the type after the fact? Or can we perhaps write a version of `diff` in terms of `take` and `__sub__`?

**Comment by jreback** [comment]
> Isn't the correct fix to to dispatch `diff` to the EA, rather than trying to fix the type after the fact? Or can we perhaps write a version of `diff` in terms of `take` and `__sub__`?

this is actually a good idea

would be simple to write as a generic EA
method

**Comment by MarcoGorelli** [comment]
@TomAugspurger @jreback I don't understand.

Currently, we have
```
        result = algorithms.diff(com.values_from_object(self), periods)
```
In the case of
```
pd.Series([1, 2, 3], dtype="Int16").diff()
```
then 
```
com.values_from_object(self)
```
is already converting the values to Object before even reaching `algorithms.diff`:
```
>>> com.values_from_object(self)
array([1, 2, 3], dtype=object)
```

So should `com.values_from_object` be changed?

Anyway, I realise that there's a somewhat tight deadline to get v1.0.0 out, and I'm happy to keep tackling it, but if you think this is (currently) beyond my capabilities, I'm happy for someone else to take over :)

**Comment by TomAugspurger** [comment]
> So should com.values_from_object be changed?

We would need to not call that, at least for EAs, but probably for all dtypes, before passing it off. I can probably take a look tomorrow.

**Comment by MarcoGorelli** [comment]
@TomAugspurger OK thanks - I *might* have figured out what you were suggesting anyway, have pushed another commit

**Comment by TomAugspurger** [comment]
See https://github.com/pandas-dev/pandas/pull/31025 for an alternative which allows EAs to define how the operation should be done.

**Comment by TomAugspurger** [comment]
I think we'll proceed with https://github.com/pandas-dev/pandas/pull/31025. Apologies for the duplicated effort @MarcoGorelli.

**Comment by MarcoGorelli** [comment]
It's OK, it was a good learning experience to attempt this and then read through your PR!

**Pull request closed without merging**